### PR TITLE
spanner custom converters applied to key components

### PR DESCRIPTION
**Pull request by ChengyuanZhao**
addresses #579

### CODE REVIEW

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConversionUtils.java [hunk 1]
```diff
@@ -128,7 +129,19 @@ public ByteArray convert(byte[] bytes) {
...
 
-	static Class boxIfNeeded(Class propertyType) {
+	/**
+	 * Converters to and from Spanner types.
+	 */
+	public static final Collection<Converter> DEFAULT_SPANNER_CONVERTERS = ImmutableSet
+			.<Converter>builder().addAll(DEFAULT_SPANNER_WRITE_CONVERTERS)
+			.addAll(DEFAULT_SPANNER_READ_CONVERTERS).build();
+
+	public static Set<Class> SPANNER_KEY_COMPATIBLE_TYPES = ImmutableSet.<Class>builder()
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConversionUtils.java]: I'm wondering why we needed to introduce this? Aren't these already in the mapping keySets anyway?

Comment (balopat) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConversionUtils.java]: Oh I see - I guess we can't have arrays for keys, that's the reason, right? It's interesting that ByteArray is supported.

Comment (balopat) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConversionUtils.java]: can we not have this as `public`? And maybe actually move it to the `MappingSpannerConverter` not a utility class? Same applies to `DEFAULT_SPANNER_CONVERTERS`

Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConversionUtils.java]: changes made

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/support/SimpleSpannerRepository.java [hunk 2]
```diff
@@ -175,7 +184,7 @@ private Object convertKeyPart(Object object) {
 		 */
 		for (Class validKeyType : MappingSpannerWriteConverter.singleItemType2ToMethodMap
 				.keySet()) {
-			if (!ConversionUtils.SPANNER_KEY_COMPATIBLE_TYPES.contains(validKeyType)) {
+			if (!SPANNER_KEY_COMPATIBLE_TYPES.contains(validKeyType)) {
```
[the remark this answers is not in the document]
Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/support/SimpleSpannerRepository.java]: done

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/support/SimpleSpannerRepository.java [hunk 3]
```diff
@@ -154,8 +157,36 @@ public Page findAll(Pageable pageable) {
...
-			k = key instanceof Key ? (Key) key : Key.of(key);
+			k = Key.class.isAssignableFrom(key.getClass()) ? (Key) key
+					: Key.of(convertKeyPart(key));
 		}
 		return operation.apply(k);
 	}
+
+	private Object convertKeyPart(Object object) {
+		if (MappingSpannerConverter
+				.isValidSpannerKeyType(ConversionUtils.boxIfNeeded(object.getClass()))) {
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/support/SimpleSpannerRepository.java]: `isValidSpannerKeyType` should be non-static, even if the logic currently depends on a set of statically predefined values. Introducing these hard-coded static depedencies makes the code hard to test and extend later, it's too rigid of a design with no particular benefit.

File: spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/support/SimpleSpannerRepository.java [hunk 4]
```diff
@@ -154,8 +157,36 @@ public Page findAll(Pageable pageable) {
...
+			return object;
+		}
+		/*
+		 * Iterate through the supported Key component types in the same order as the
+		 * write converter. For example, if a type can be converted to both String and
+		 * Double, we want both the this key conversion and the write converter to choose
+		 * the same.
+		 */
+		for (Class validKeyType : MappingSpannerWriteConverter.singleItemType2ToMethodMap
```
Comment (balopat) [inline spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/support/SimpleSpannerRepository.java]: same thing here - I would rather hide the detail of `singleItemType2ToMethodMap.keySet()` and expose a non-static member function `writableTypes()`

Comment (ChengyuanZhao) [reply spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/repository/support/SimpleSpannerRepository.java]: changes made

**Pull request merged**